File: source/net/auoeke/csf/parser/Context.java
package net.auoeke.csf.parser;

import net.auoeke.csf.parser.lexer.lexeme.Token;

public enum Context {
    FILE,
    ARRAY,
    MAP;

    public Token begin() {
        return switch (this) {
            case ARRAY -> Token.ARRAY_BEGIN;
            case MAP -> Token.MAP_BEGIN;
            case FILE -> null;
        };
    }

    public Token end() {
        return switch (this) {
            case ARRAY -> Token.ARRAY_END;
            case MAP -> Token.MAP_END;
            case FILE -> null;
        };
    }

    public String description() {
        return switch (this) {
            case ARRAY -> "an array";
            case MAP -> "a map";
            case FILE -> "a file";
        };
    }

    public static Context of(char character) {
        return switch (character) {
            case '[' -> ARRAY;
            case '{' -> MAP;
            default -> throw new Error();
        };
    }
}

File: src/net/auoeke/cin/parser/Context.kt
package net.auoeke.cin.parser

import kotlin.contracts.contract

enum class Context {
    FILE,
    ARRAY,
    MAP;

    val start: Char get() = when (this) {
        ARRAY -> '['
        MAP -> '{'
        FILE -> Char.MIN_VALUE
    }

    val description: String get() = when (this) {
            ARRAY -> "an array"
            MAP -> "a map"
            FILE -> "a file"
        }

    companion object {
        val Char.context: Context get() = when (this) {
            '[' -> ARRAY
            '{' -> MAP
            else -> throw Error()
        }
    }
}
